app/frontend/src/store/modules/ches.js

```diff
@@ -71,38 +84,45 @@ export default {
       commit('ADD_TXID', txId);
     },
 
-    addTableData({ commit }, data) {
-      commit('ADD_TABLEDATA', data);
-    },
-    // commit,
-    populateTable({ commit, state }) {
-      try {
-        // for each existing tx in txIds array
-        state.txIds.forEach(async (tx) => {
-          // if tableData doesnt contain an object with txId = tx.txId
-          if (!state.tableData.find((o) => o.txId === tx.txId)) {
-            // for each of the messages in tx
-            tx.messages.forEach(async (message) => {
-              // get details from CHES and add to tableData
+    // populates history table each time the component is mounted
+    async populateTable({ commit, state, dispatch }) {
+
+      commit('UPDATE_TABLELOADING', true);
+
+      // for each of the transactions that exists in local storage
+      state.txIds.forEach( (tx) => {
+        // for each of the messages in tx
+        tx.messages.forEach(async (message) => {
+          // if existing tableData doesnt contain this transaction, and the message in this row of table has a status that is not 'completed' or 'cancelled'
+          if (!state.tableData.find((o) => (o.txId === tx.txId && (o.status === 'completed' || o.status === 'cancelled')))) {
```

Comment:
Looking at this logic, I "think" you can probably lift this if block up to the first forEach loop level as the conditional doesn't depend on the internal message object at all. By lifting it up, you can also reduce the amount of computation required (only act if transactions and statuses match instead of doing the same check for every message).
```js
const statuses = ['completed', 'cancelled'];
state.txIds.forEach(tx => {
  if (!state.tableData.find(o => (o.txId === tx.txId && statuses.some(status => o.status.includes(status)) {
    tx.messages.forEach(async (message) => {
```
May need to have a quick chat on this.

Reply:
@jujaga As I think you suggested, i have changed this code. i'm now flattenning the nested array of txIds to become just an array of messages. and looping through those while checking if tableData has matching row or status that isnt terminal.
I believe we're good to merge this now..